### PR TITLE
sql: create a span per sql statement

### PR DESCRIPTION
**Pull request by andreimatei**
Before this patch, statements did not have their own spans. Instead,
transactions had spans. This dates from a time when the TxnCoordSender
would capture the span of the first write and expect it to live for the
duration of the transaction, but that stopped being the case a while
ago.
This patch makes each statement create its own span.

Release note: None

### CODE REVIEW

**Review by jordanlewis — COMMENTED** [review]
Generally seems like a good idea, though for better or for worse changes like this always make me nervous about performance. A few comments.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-LaIc3Uh-cFwpRJTVQqZ:b-e1r8r8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/sql/conn_executor.go, line 1155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-LaIb76s46C18NYa21qP:-LaIb76t2YaFZRiTlPN2:bmj6bts) ([raw file](https://github.com/cockroachdb/cockroach/blob/484394c6cb6167b35a2cab88eb405b8cdea75b78/pkg/sql/conn_executor.go#L1155)):*
> ```Go
> // drainingComplete is returned by execCmd when the connExecutor previously got
> // a DrainRequest and the time is ripe to finish this session.
> var drainingComplete error = fmt.Errorf("this is a good time to finish this session")
> ```

This doesn't ever seem to get returned?
___
*[pkg/sql/conn_executor.go, line 1168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-LaIbSpK8_avnBmez_Gs:-LaIbSpK8_avnBmez_Gt:bh0j0kx) ([raw file](https://github.com/cockroachdb/cockroach/blob/484394c6cb6167b35a2cab88eb405b8cdea75b78/pkg/sql/conn_executor.go#L1168)):*
> ```Go
> // draining is complete.
> // The draining return value is true if the session entered draining mode
> // (because we received a DrainRequest) or if the session continues draining in
> ```

I would appreciate a more verbose description of "draining" here as we use this word in several different contexts throughout the codebase. Also, I'm confused why we're also passing in a value for draining as well as returning one. If this is a state variable, should it live on the connExecutor instead?

**Review by jordanlewis — COMMENTED** [review]
Also to be clear this would have my LGTM if it wasn't so late in the release cycle. As it is it seems late. For master LGTM, for backport... I don't know.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-LaIcLOzCO2LYrnBtN8o:bj3mko6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/conn_executor.go, line 1155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-LaIb76s46C18NYa21qP:-LaqW4Ak-rD6yWv832fR:b-mht0cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/484394c6cb6167b35a2cab88eb405b8cdea75b78/pkg/sql/conn_executor.go#L1155)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This doesn't ever seem to get returned?
</blockquote></details>

I think I was hungry when typing this, dunno what I did. See now.
___
*[pkg/sql/conn_executor.go, line 1168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-LaIbSpK8_avnBmez_Gs:-LaqWAjM-_S7se4Kwq3P:b-uhplad) ([raw file](https://github.com/cockroachdb/cockroach/blob/484394c6cb6167b35a2cab88eb405b8cdea75b78/pkg/sql/conn_executor.go#L1168)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I would appreciate a more verbose description of "draining" here as we use this word in several different contexts throughout the codebase. Also, I'm confused why we're also passing in a value for draining as well as returning one. If this is a state variable, should it live on the connExecutor instead?
</blockquote></details>

moved to the connEx as a field.

Added a couple more words to a couple of comments. See if you're satisfied.

**Review by jordanlewis — COMMENTED** [review]
:lgtm:

As I'm sure we've discussed, I wish there was a way in Go to have a static declaration of the kinds of errors a function could return. `execCmd` is now special because it isn't designed to permit the `if err != nil { return err }` semantic - it requires a special check and lack of propagation. But there's no way you can know that unless you read the comments and code. Nothing to do I don't think, as you've commented, just complaining about it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-LaqZZGQBzvWtDpZ3tva:b-rxw008)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

**Review by andreimatei — COMMENTED** [review]
well actually it'd be pretty OK for `ex.run()` to just return those errors, because the return val of `ex.run()` gets pretty much ignored later on - the connection is terminated and that's pretty much it regardless of whether `ex.run()` declares "success" or "failure" - and I'm making that be even more true in a separate PR.
That, and the fact that we were already handling `io.EOF` as special is why I didn't try hard to make a more explicit interface.

Btw, I had ran some of our benchmarks on this and they showed no difference.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-LaqdYiPBfD7xhr1cRz7:b-ivirvv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis)

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-Latnaih8xpGZZ4d9kNP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @jordanlewis)

___
*[pkg/sql/conn_executor.go, line 1163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-Latmywb5KYW9xW6z5BL:-Latmywb5KYW9xW6z5BM:b-fap65m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9d3c2be2e3b935ceeaecf1805df2bb88f640d6/pkg/sql/conn_executor.go#L1163)):*
> ```Go
> // a DrainRequest and the time is ripe to finish this session (i.e. we're no
> // longer in a transaction).
> var errDrainingComplete = fmt.Errorf("this is a good time to finish this session")
> ```

[nit] this should mention "drain request"
___
*[pkg/sql/conn_executor.go, line 1175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-LatnKw77c9VHYsERKvj:-LatnKw77c9VHYsERKvk:b9bp6tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9d3c2be2e3b935ceeaecf1805df2bb88f640d6/pkg/sql/conn_executor.go#L1175)):*
> ```Go
> // connection is found to be idle).
> func (ex *connExecutor) execCmd(ctx context.Context) error {
> 	ctx, sp := tracing.EnsureChildSpan(ctx, ex.server.cfg.AmbientCtx.Tracer, "exec cmd")
> ```

It would be helpful if the span title contained the type of command (at least special cases for "stmt" and "prepare")

**Review by andreimatei — COMMENTED** [review]
TFTRs

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-LbPwMOn7bNHLL5XCYui:b-6rbzir)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jordanlewis)

___
*[pkg/sql/conn_executor.go, line 1163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-Latmywb5KYW9xW6z5BL:-LbPYHSv97khmVM12-QY:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9d3c2be2e3b935ceeaecf1805df2bb88f640d6/pkg/sql/conn_executor.go#L1163)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this should mention "drain request"
</blockquote></details>

done
___
*[pkg/sql/conn_executor.go, line 1175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-LatnKw77c9VHYsERKvj:-LbPYuIE963_jN8VQ9Gg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9d3c2be2e3b935ceeaecf1805df2bb88f640d6/pkg/sql/conn_executor.go#L1175)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It would be helpful if the span title contained the type of command (at least special cases for "stmt" and "prepare")
</blockquote></details>

done

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35805#-:-LbPxSDv3IrVoPeyRjsf:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jordanlewis)

**Pull request merged**
